[정승호] 작품 상세 페이지 엔드포인트(feature/work_detail)

seungho-jeong opened the pull request:
## 수정 사항 간략한 한줄 요약
#### Commit 1
- 작품(Artwork) 상세 페이지 API 구현
#### Commit 2
- WorkDetailView 부분 수정
#### Commit 3
- 댓글 작성/관리 및 좋아요 기능 구현
		 
## 수정 사항들 자세한 내용
#### Commit 1
- 작품 상세정보 View 작성(WorkDetailView)
- 댓글 관리(생성, 삭제, 수정) View 작성(CommentView)
#### Commit 2
- Comments filtering 코드 수정(Query 수행 => 미리 지정한 Local 변수 사용)
- Comments objects를 count하는 코드 수정(Python's len() => Django's count())
#### Commit 3
- 댓글 POST, PATCH, DELETE 및 댓글 좋아요 기능 구현
Token을 활용한 User 식별 기능은 추가 중
- 작품 평가(좋아요) 기능 구현
평가는 유저가 게시물당 한 번만 평가 가능함
=> Ex 1 : 이미 "좋아요"를 누른 게시물에 "좋아요"를 한 번 더 누르면 좋아요가 취소됨(좋아요 1 감소)
=> Ex 2 : "좋아요"를 누른 게시물에 "감동 받았어요"를 누르면 "좋아요"가 취소(좋아요 1 감소)되고 "감동 받았어요"가 1 증가함

## 기타 질문 및 특이 사항
- 상세 페이지를 불러올 때 댓글 내용을 한꺼번에 불러오도록 만들었고 댓글 생성, 수정, 삭제 시에도 댓글 리스트를 반환하도록 만들었습니다.
이 방식이 왠지 댓글을 여러번 불러와서 중복되는 느낌이 드는데 어떤 방법이 합리적일까요?

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

seungho-jeong committed:
Add: 작품(Artwork) 상세 페이지 API 구현
- 작품 상세정보 View 작성(WorkDetailView)
- 댓글 CRUD View 작성(CommentView)

soheon-lee commented on work/views.py:
```diff
@@ -0,0 +1,125 @@
+            this_work     = Work.objects.get(id = work_id)
+            creator       = this_work.user
+            related_works = creator.work_set.exclude(id = work_id)
+            comments      = Comment.objects.filter(work_id = work_id)
```
work 객체를 이미 13번째 줄에서 특정해주셨기 때문에, work와 comment 가 연결되어 있음을 활용해 work의 속성으로 접근해주시는 것이 좋습니다.

seungho-jeong replied:
적용하였습니다. 어제 반영했는데 댓글을 못 남겼네요 ㅠㅠ

soheon-lee commented on work/views.py:
```diff
@@ -0,0 +1,125 @@
+                "views"       : this_work.views,
+                "created_at"  : this_work.created_at,
+                "updated_at"  : this_work.updated_at,
+                "image_url"   : [ workimage.image_url for workimage in this_work.workimage_set.all() ],
+                "tag"         : [ tag.name for tag in this_work.tag.all() ],
```
gooood!

soheon-lee commented on work/views.py:
```diff
@@ -0,0 +1,125 @@
+                "updated_at"  : this_work.updated_at,
+                "image_url"   : [ workimage.image_url for workimage in this_work.workimage_set.all() ],
+                "tag"         : [ tag.name for tag in this_work.tag.all() ],
+                "commentNum"  : len(comments),
```
객체의 수를 세어주는 django의 메소드를 활용해보세요 :)

seungho-jeong replied:
len대신 Django의 count를 사용하였습니다.

soheon-lee commented on work/views.py:
```diff
@@ -0,0 +1,125 @@
+        try:
+            if Work.objects.filter(id = work_id).exists():
+
+                user_name       = data["user_name"]
```
user 정보는 데코레이터에서 받아오시면 더 좋을 것 같습니다 :)

seungho-jeong replied:
데코레이터를 구현할 예정이지만 시간이 많이 소요될 것 같아 기초 기능 구현 후 하려고 계획하였습니다.
우선은 이 형태로 하겠습니다 ㅠㅠ

seungho-jeong committed:
Modify: work/views.py/WorkDetailView 부분 수정
- Comments filtering하는 코드 수정(Query 수행 -> 미리 지정한 변수 재사용)
- Comments OBJ를 Count하는 코드 수정(python's len() => Django's count())

seungho-jeong committed:
Add: 댓글 작성/관리 및 좋아요 기능 구현
- 댓글 POST, PATCH, DELETE 및 댓글 좋아요 기능 구현
* Token을 활용한 User 식별 기능은 추가 중
- 작품 평가(좋아요) 기능 구현
* 평가는 유저가 게시물당 한 번만 평가 가능함

lordmyshepherd commented on work/views.py:
```diff
@@ -0,0 +1,178 @@
+            return JsonResponse({"MESSAGE": "DELETE_SUCCESS"}, status=200)
+
+        except Comment.DoesNotExist:
+            return JsonResponse({"MESSAGE": "DOES_NOT_EXIST_COMMENT"}, status=400)
```
```
try : 
    comment = Comment.objects.get(id = comment_id, work_id = work_id)
    comment.delete()
except Comment.DoesNotExist:
    return JsonResponse....
    
```

lordmyshepherd replied:
코드 참고해서 프로젝트 끝나고 수정해보시면 좋을 것 같아요!

seungho-jeong replied:
반영하였습니다!

seungho-jeong committed:
Add : 댓글 관리기능/좋아요 기능 수정 및 배경화면 상세 페이지 API 구현

seungho-jeong committed:
Add : Decorator 추가(user/utils.py)

seungho-jeong committed:
Modify: 컨플릭트 발생 부분 수정

soheon-lee commented on user/views.py:
```diff
@@ -30,44 +30,50 @@ def post(self, request):
+                    introduction=data["introduction"],
+                    profile_image_url=data["profile_image_url"]
+                )
+                return JsonResponse({"MESSAGE": "SUCCESS"}, status=201)
```
else가 없어도 바로 위의 if문이 만족되지 않는다면 바로 도달할 수 있는 로직입니다. 꼭 필요한 경우가 아니면 else를 사용하지 않는 것이 보다 파이썬다운 코딩 스타일입니다. 지워주셔도 됩니다 :)

soheon-lee commented on user/views.py:
```diff
@@ -30,44 +30,50 @@ def post(self, request):
                 )
-                return JsonResponse({"MESSAGE": "LOGIN_SUCCESS", "token": token.decode("UTF-8")}, status=200)
+                return JsonResponse({"MESSAGE": "LOGIN_SUCCESS", "Authorization": token.decode("UTF-8")}, status=200)
 
             else:
                 return JsonResponse({"MESSAGE": "INVALID_INPUT"}, status=401)
```
else가 없어도 바로 위의 if문이 만족되지 않는다면 바로 도달할 수 있는 로직입니다. 꼭 필요한 경우가 아니면 else를 사용하지 않는 것이 보다 파이썬다운 코딩 스타일입니다. 지워주셔도 됩니다 :)

soheon-lee commented on work/views.py:
```diff
@@ -22,8 +25,223 @@
+        this_work.views += 1
+        this_work.save()
+
+        comments      = Comment.objects.filter(work_id = work_id).prefetch_related("work")
+        related_works = Work.objects.filter(user_id = this_work.user.id).exclude(id = work_id)
```
31번에서 선언하신 `this_work` 객체에서 출발해서 접근하시면 Comment 클래스 Work 클래스에 다시 접근해서 쿼리를 날리는 일을 줄일 수 있습니다.

soheon-lee commented on work/views.py:
```diff
@@ -22,8 +25,223 @@
+                "creator_id"  : this_work.user.id,
+                "creator_img" : this_work.user.profile_image_url,
+                "user_id"     : request.user.id if request.user else "GEUST",
+                "user_name"   : User.objects.get(id=request.user.id).user_name if request.user else "GEUST",
+                "user_image"  : User.objects.get(id=request.user.id).profile_image_url if request.user else "GEUST",
```
```suggestion
                "user_name"   : request.user.user_name if request.user else "GEUST",
                "user_image"  : request.uesr.profile_image_url if request.user else "GEUST",
```

soheon-lee commented on work/views.py:
```diff
@@ -22,8 +25,223 @@
+                "image_url"   : [ workimage.image_url for workimage in this_work.workimage_set.all() ],
+                "tag"         : [ tag.name for tag in this_work.tag.all() ],
+                "commentNum"  : comments.count(),
+                "likeBtnNum"  : LikeIt.objects.filter(work_id = work_id).count(),
+                "followerNum" : Follow.objects.filter(creator_id = this_work.user.id).count(),
+                "followingNum": Follow.objects.filter(follower_id = this_work.user.id).count(),
```
31번에서 선언하신 this_work 객체에서 출발해서 접근하시면 Follow 클래스에 다시 접근해서 쿼리를 날리는 일을 줄일 수 있습니다.

seungho-jeong committed:
Modify: user/views.py 불필요 Else문 삭제, work/views.py 쿼리 간소화 등

lordmyshepherd commented on work/views.py:
```diff
@@ -22,8 +25,222 @@
+class WallpaperDetailView(View):
+    def get(self, request, wallpaper_id):
+        try:
+            this_wallpaper = WallpaperImage.objects.select_related("work").get(id = wallpaper_id)
```
```suggestion
            this_wallpaper = WallpaperImage.objects.select_related("work", "work__user").get(id = wallpaper_id)
```

lordmyshepherd commented:
LGTM 👍
